### ISSUE TITLE
Support for condition-level `observedGeneration` wait rule

### ISSUE THREAD

**Issue by dpb587** [k14s/kapp#435]
**Describe the problem/challenge you have**

Hi - I'm trying to configure a resource wait rule including `observedGeneration` from the condition-level, but I don't think it's possible. I'd be interested if there is an alternative, but seems like this might be a feature request from looking into it more.

For a specific scenario, I'm using [cert-manager](https://cert-manager.io) and have a custom wait rule for `Certificate`. The wait rule works fine for new resources, but if an existing certificate gets updated (e.g. its `dnsNames` changes) it does not wait since `Ready` is already present and true. I tried the [`supportsObservedGeneration` flag](https://carvel.dev/kapp/docs/v0.45.0/config/#wait-rules), but it didn't work (i.e. ongoing "waiting for generation * to be observed"). Digging into the code I found kapp [only checks `status.observedGeneration`](https://github.com/vmware-tanzu/carvel-kapp/blob/e0bf1c9d485c5c59a23b4cfe75915782e7fb4b42/pkg/kapp/resourcesmisc/custom_waiting_resource.go#L52-L55), not `status.conditions[*].observedGeneration` (which is what [cert-manager provides](https://github.com/cert-manager/cert-manager/blob/b1180c59ad588e73ac25b0d70a86661cf7c180e1/pkg/api/util/conditions.go#L176), [not status-level](https://github.com/cert-manager/cert-manager/blob/ad4264b6ecd48702b4a3b83b06bd00bd32a19df5/deploy/crds/crd-certificates.yaml#L297-L359)).

It looks like [KEP-1623: Standardize Conditions](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions#proposal) formalized `observedGeneration` as a [condition-level field](https://github.com/kubernetes/apimachinery/blob/14143357d417e6d4c3c3669265d13f41d6eb8adb/pkg/apis/meta/v1/types.go#L1483-L1488), so it seems like cert-manager is following standards, at least. That KEP was released in v1.19 (August 2020) which was after the original [kapp pull request](https://github.com/vmware-tanzu/carvel-kapp/pull/115) that introduced these custom wait rules, so I'd guess the condition-level behavior just hasn't been previously considered here. It also just seems much less common as I haven't found many other projects including the condition field in their schema ([contour does](https://github.com/vmware-tanzu/community-edition/blob/3f967483a08016aecba2901f7c4a4152387474a7/addons/packages/contour/1.18.2/bundle/config/upstream/contour.yaml#L477)), but I do see that `kubectl wait` officially respects it when present according to [tests](https://github.com/kubernetes/kubernetes/blob/4033355d292505eaed05655bd91f4ea9edc7ef27/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go#L926-L1000).

<details>
<summary><code>kapp/Config</code> debugging</summary>

For reference, I was using the following config when trying to get this to work.

```yaml
apiVersion: kapp.k14s.io/v1alpha1
kind: Config
waitRules:
- resourceMatchers:
  - apiVersionKindMatcher:
      apiVersion: cert-manager.io/v1
      kind: Certificate
  # hangs; needs to look at .status.conditions[*].observedGeneration instead
  # supportsObservedGeneration: true
  conditionMatchers:
  - type: Ready
    status: "True"
    success: true
```

</details>

**Describe the solution you'd like**

Assuming I haven't misunderstood something, I think kapp should add support for `supportsObservedGeneration` at the condition matcher-level. I wouldn't expect operators to typically need it for both the wait rule and condition matcher, but it would technically be possible.

For the example scenario, I'd then use the following.

```yaml
apiVersion: kapp.k14s.io/v1alpha1
kind: Config
waitRules:
- resourceMatchers:
  - apiVersionKindMatcher:
      apiVersion: cert-manager.io/v1
      kind: Certificate
  conditionMatchers:
  - type: Ready
    status: "True"
    success: true
    supportsObservedGeneration: true
```

Let me know if there are alternatives or this sounds reasonable - I could probably provide a small PR with the change.

Thanks!

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by praveenrewar** [comment]
Hi @dpb587, thanks for bringing this to our notice, this seems like a reasonable ask. 

> I could probably provide a small PR with the change.

We would be happy to review it on priority.

**Comment by kumaritanushree** [comment]
[dpb587](https://github.com/dpb587), would you be able to a submit a PR for this? If you don't have cycles to take this up, let us know. we could take it up

**Comment by dpb587** [comment]
I don't see myself getting back to this in the short-term as I originally planned, though do still look forward to the change. Feel free to pick this up if you/others are able.

**Comment by kumaritanushree** [comment]
Acceptance Criteria:

- Users can define wait rule for status.condition level 
**Reference(tested with CR cert-manager's certificate https://cert-manager.io/docs/usage/certificate/)**
_I have used kapp config while deploying certificate :_
```
apiVersion:` kapp.k14s.io/v1alpha1
kind: Config
waitRules:
- resourceMatchers:
  - apiVersionKindMatcher:
      apiVersion: cert-manager.io/v1
      kind: Certificate
  conditionMatchers:
  - type: Ready
    status: "False"
    success: true
    supportsObservedGeneration: true
  ```
    
In this scenario what kapp will do is, kapp will check if conditionMatchers.supportsObservedGeneration = true then it will match Metadata.Generation with Status.Condition.ObservedGeneration , also check for type =ready and status=false then kapp will say the resource got deployed successfully.

**Comment by dpb587** [comment]
Thanks, @kumaritanushree and team!

**Comment by renuy** [comment]
Fix available as part of kapp v0.47.0
